### PR TITLE
Introduce Ruff

### PR DESCRIPTION
**Pull request by glatterf42**
# Please confirm that this PR has done the following:

- ~[ ] Tests Added~ But updated black CI workflow
- [x] Documentation Added
- [x] Name of contributors Added to AUTHORS.rst
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

As suggested by https://github.com/IAMconsortium/pyam/pull/820#issuecomment-1971061133, this PR cherry-picks the commits from #820 that refer to ruff and amends some of them that got mixed with Python mentions.

### CODE REVIEW

**Comment by glatterf42** [comment]
Should we add some mention of ruff to the docs? 

Also, in ixmp4, we have an optional dependency group called "dev" in pyproject.toml and that includes a minimum version for ruff. Should we add something similar here?

**Comment by danielhuppmann** [comment]
> Should we add some mention of ruff to the docs?

Please also add the badge to index.rst, should be sufficient.

> Also, in ixmp4, we have an optional dependency group called "dev" in pyproject.toml and that includes a minimum version for ruff. Should we add something similar here?

Yes, please.

File: pyproject.toml
Comment (phackstock) [inline pyproject.toml:1]: Just so that I understand it correctly, we will now have a `setup.cfg`, `setup.py` **and** `pyproject.toml`?
Could you combine everything into a pyproject.toml or move the ruff settings to setup.cfg or setup.py?

File: tests/test_feature_aggregate.py
Comment (phackstock) [inline tests/test_feature_aggregate.py:404]: It seems that this can actually be removed, if it was commented out and the test did not break.
```suggestion
```

**Review by phackstock — CHANGES_REQUESTED** [review]
Looks quite good overall, a few small comments in line. Maybe you could mention `ruff` the `PULL_REQUEST_TEMPLATE.md` or `CONTRIBUTING.rst`

File: pyproject.toml
Comment (glatterf42) [reply pyproject.toml:1]: Yes, ruff won't read configuration from `setup.cfg` or `setup.py`. My intention with using `pyproject.toml` was to next use #823 to migrate to poetry, which will read all settings from `pyproject.toml`, so that way the files will be combined. Let me clean up this PR as per your suggestions and then create one with poetry and then we can decide: if you don't like poetry, I can migrate the settings here to `ruff.toml`.

File: tests/test_feature_aggregate.py
Comment (glatterf42) [reply tests/test_feature_aggregate.py:404]: I just wasn't sure if this was kept in for some potential new test in the future, but sure, I can delete it.

Comment (danielhuppmann) [reply tests/test_feature_aggregate.py:404]: I suggest leave but uncomment  and use (see other suggestion) for consistency 
```suggestion
    w = "Emissions|CO2"
```

File: tests/test_feature_aggregate.py
Comment (danielhuppmann) [inline tests/test_feature_aggregate.py:407]:
```suggestion
    pytest.raises(ValueError, simple_df.aggregate_region, v, weight=w, method="max")
```

Comment (glatterf42) [reply tests/test_feature_aggregate.py:407]: I can add this as a test case rather than removing `w` :)

File: docs/index.rst
Comment (danielhuppmann) [inline docs/index.rst:24]: Please remove the Black badge

Comment (glatterf42) [reply docs/index.rst]: Sorry, forgot that.

**Comment by danielhuppmann** [comment]
You also have to use the ruff-badge in docs/index.rst on line 8.

**Review by danielhuppmann — APPROVED** [review]
Looks good to me, thank you!

File: pyproject.toml
Comment (phackstock) [reply pyproject.toml:1]: I see your point. I'd prefer though if you use `ruff.toml` for now and then in a future PR integrate `setup.cfg` and `setup.py` into one `pyproject.toml`.

**Review by phackstock — APPROVED** [review]
Good to merge from my side if you change the pyproject.toml to ruff.toml

**Comment by glatterf42** [comment]
I can't merge it myself though, don't have permissions for that, so please go ahead.

**Pull request merged**
